saw-core/prelude/Prelude.sawcore

```diff
@@ -1389,6 +1474,287 @@ primitive mkDouble : Integer -> Integer -> Float;
+         (Nothing (EqP Bool False True));
+
+-- Compare two bitvectors with bvule, returning a proof if bvule succeeds
+bvuleWithProof : (n : Nat) -> (v1 v2 : bitvector n) ->
+                 Maybe (EqP Bool (bvule n v1 v2) True);
+bvuleWithProof n v1 v2 =
+  iteDep (\ (b:Bool) -> Maybe (EqP Bool b True)) (bvule n v1 v2)
+         (Just (EqP Bool True True) (ReflP Bool True))
+         (Nothing (EqP Bool False True));
+
+-- Convert a proof of bitvector equality to one of Nat equality
+primitive bvEqToEqNat : (n : Nat) -> (v1 v2 : bitvector n) ->
+                        EqP (bitvector n) v1 v2 ->
+                        eqNat (bvToNat n v1) (bvToNat n v2);
+
+-- Convert a proof of bitvector less-than to one of Nat less-than
+primitive bvultToIsLtNat : (n : Nat) -> (v1 v2 : bitvector n) ->
+                           EqP Bool (bvult n v1 v2) True ->
+                           IsLtNat (bvToNat n v1) (bvToNat n v2);
+
+-- | Index a vector using a proof that the index is in the range of the vector
+-- FIXME: atWithDefault should maybe use this...?
+primitive atWithProof : (n : Nat) -> (a : sort 0) -> a -> Vec n a ->
+                        (i : Nat) -> IsLtNat i n -> a;
+
+-- Set the value at index i in a vector using a proof that i is in range
+primitive updWithProof : (n : Nat) -> (a : sort 0) -> Vec n a ->
+                         (i : Nat) -> a -> IsLtNat i n -> Vec n a;
+
+-- Take a slice of a vector using a proof that the slice is in range
+primitive sliceWithProof : (a : sort 0) -> (n off len : Nat) ->
+                           IsLeNat (addNat off len) n -> Vec n a -> Vec len a;
+
+-- Update a slice of a vector using a proof that the slice is in range
+primitive updSliceWithProof : (a : sort 0) -> (n off len : Nat) ->
+                              IsLeNat (addNat off len) n ->
+                              Vec n a -> Vec len a -> Vec n a;
+
+
+--------------------------------------------------------------------------------
+-- Vectors indexed by bitvectors
+
+-- Helper definition to write the proposition that x <u y
+is_bvult : (n:Nat) -> (x y:bitvector n) -> Prop;
+is_bvult n x y = EqP Bool (bvult n x y) True;
+
+-- Helper definition to write the proposition that x <=u y
+is_bvule : (n:Nat) -> (x y:bitvector n) -> Prop;
+is_bvule n x y = EqP Bool (bvule n x y) True;
+
+-- Axiom: x <u 0 never hold
+axiom not_bvult_zero : (n:Nat) -> (x:bitvector n) ->
+                       EqP Bool (bvult n x (bvNat n 0)) False;
+
+-- Axiom: x <=u y <u z implies x <u z
+axiom trans_bvult_bvule : (n:Nat) -> (x y z:bitvector n) ->
+                          is_bvult n x y -> is_bvule n y z -> is_bvult n x z;
+
+axiom bvult_sub_add_bvult : (n:Nat) -> (x y z:bitvector n) ->
+                            is_bvule n y z -> is_bvult n x (bvSub n z y) ->
+                            is_bvult n (bvAdd n y x) z;
+
+-- The type of a bitvector-indexed array of length len : bitvector n
+primitive BVVec : (n:Nat) -> bitvector n -> sort 0 -> sort 0;
```

Comment:
I'm puzzled about why this type needs to exist as a primitive. Couldn't you just define it as:
```
BVVec n w a = Vec (bvToNat n w) a
```

Reply:
This certainly makes sense. My one concern is whether there are any definitional equalities that no longer work when defining `BVVec` this way. I think probably not, but I will have to try it out. Maybe I'll add a Heapster ticket to investigate this change, and if it works I'll let you know.